### PR TITLE
Add tuples

### LINKED ISSUE #21: Tuples
We use tuples in a couple places and they're very much like immutable ordered sets. The syntax is typically (element1, element2).

### PR DESCRIPTION
**Pull request by annevk**
Fixes #21.

***
[Preview](https://infra.spec.whatwg.org/branch-snapshots/annevk/tuples/) | [Diff](https://s3.amazonaws.com/pr-preview/whatwg/infra/f4d2de2...865ee66.html)

### CODE REVIEW

**Comment by annevk** [comment]
This isn't quite right.

File: infra.bs [hunk 1]
```diff
@@ -547,6 +547,27 @@ ordered sets; implementations can optimize based on the fact that the order is n
...
+
+<p>Some lists are designated as <dfn export lt=tuple>tuples</dfn>. A tuple is a <a>list</a> with the
+additional semantic that its <a for=list>size</a> must not change and that its
+<a for=tuple>items</a> can be indexed by <dfn export for=tuple>name</dfn>.
+
+<p>For notational convenience, a literal syntax can be used to express <a>tuples</a>, by surrounding
+the tuple with ( ) characters and separating the <a for=tuple>items</a> by comma. The
+<a for=tuple>items</a> can be preceded by <a for=tuple>name</a> and : character if the
+<a for=tuple>names</a> are not clear from context. An indexing syntax can be used by providing a
+zero-based index and named index into a list inside square brackets.
```
Comment (mikewest) [inline infra.bs]: Is a 0-based index something we need to support for tuples? Seems like they'd always be named.

Comment (annevk) [inline infra.bs]: Yeah, I guess you're right.

Your pair example is also interesting. I wonder if we should have dedicated support for those. There's numerous places that use the / syntax that way.

**Comment by annevk** [comment]
A separate type seems fine. I just found myself repeating a bunch of the list language so I used that as a shortcut. I didn't extend any of the list accessers apart from item with for=tuple though.

I like your idea of using |status|'s x instead.

And yes, I think we should have dedicated pairs with their own syntax. I'll have another go on Monday I suppose.

**Comment by annevk** [comment]
Does having pairs mean we have to refactor maps somehow? It uses the pair language already...

File: infra.bs [hunk 2]
```diff
@@ -547,6 +548,38 @@ ordered sets; implementations can optimize based on the fact that the order is n
 <a for=list>prepend</a> operation otherwise.
 
 
+<h3 id=tuples>Tuples</h3>
+
+<p>A <dfn export>tuple</dfn> is a specification type consisting of a finite ordered
+immutable-sequence of <dfn for=tuple,pair>items</dfn>, each of which has a
```
Comment (mikewest) [inline infra.bs]: What is "immutable" about the tuple? The list of names? The values associated with those names?

File: infra.bs [hunk 3]
```diff
@@ -547,6 +548,38 @@ ordered sets; implementations can optimize based on the fact that the order is n
 <a for=list>prepend</a> operation otherwise.
 
 
+<h3 id=tuples>Tuples</h3>
+
+<p>A <dfn export>tuple</dfn> is a specification type consisting of a finite ordered
```
Comment (mikewest) [inline infra.bs]: Is "specification type" defined somewhere?

File: infra.bs [hunk 4]
```diff
@@ -547,6 +548,38 @@ ordered sets; implementations can optimize based on the fact that the order is n
...
+immutable-sequence of <dfn for=tuple,pair>items</dfn>, each of which has a
+<dfn export for=tuple,pair>name</dfn>.
+
+<p>For notational convenience, a literal syntax can be used to express <a>tuples</a>, by surrounding
+the tuple with ( ) characters and separating the <a for=tuple>items</a> by comma. The
```
Comment (mikewest) [inline infra.bs]: It seems strange that we'd end up with "`a/b`" for a pair, but `(a, b, c)` for a triple+. I guess that's consistent with existing usage (at least in Fetch), but maybe it's worth changing to `(a, b)` for everything?

File: infra.bs (hunk 2 above)
Comment (annevk) [inline infra.bs]: The sequence is. So you cannot change the size.

File: infra.bs (hunk 3 above)
Comment (annevk) [inline infra.bs]: No, @domenic borrowed it from ECMAScript I think. We might want to formalize that too.

File: infra.bs (hunk 4 above)
Comment (annevk) [inline infra.bs]: That seems okay to me too. I don't mind changing Fetch around too much. It just seemed that a/b might be nice to have.

**Comment by domenic** [comment]
I changed a bunch of editorial things; let me know what you think.

I don't think maps should be based on pairs. The difference is subtle but tuples/pairs are for defining new novel specification types. So I don't think "pair" is a specification type that map would use. Rather, "status" is a specification type that fetch would define and use. There is no generic "pair" type.

Maybe we can phrase the intro better in that light. Hmm, not sure how though.

**Comment by annevk** [comment]
Added another commit. I think this ends up fixing #21 and #81, but not sure about #13 due to ordering and syntax. (Immutable) records still seem better there.

**Comment by domenic** [comment]
I think we should work to make this fix #13 before merging. I don't want two separate primitives.

Maybe we should flip things. We'd change the name to record, or maybe struct. We'd say that by default things are named, and that you can omit the names if there is a strong convention for the order that you plan to always follow.

**Comment by annevk** [comment]
Would you be okay with having structs and ordered structs, and only the later get to have syntax (and be pairs)?

**Comment by domenic** [comment]
That sounds pretty good. Editorially, I'd do that as a separate sentence under the same heading, instead of a separate heading, similar to how I changed pairs to be. That's because IMO it is good to intentionally blur the lines here and say they're all basically the same thing.

**Comment by annevk** [comment]
Done. I removed the name syntax from tuples as well since we haven't actually needed that anywhere thus far.

**Comment by domenic** [comment]
Made some small tweaks. LGTM if you like those, with an updated commit message of course.

**Comment by annevk** [comment]
@mikewest you're still on board?

**Pull request merged**